File: 2020/2/main.py
import os
import time

def firstproblem(payload):
    valid = 0
    lines = [x for x in payload.split('\n') if x]

    for line in lines:
        info = line.split(' ')
        numbers = info[0].split('-')
        min = int(numbers[0])
        max = int(numbers[1])

        letter = info[1][:len(':')]
        count = 0
        for c in info[2]:
            if c == letter:
                count+=1

        if count<=max and count>=min:
            valid+=1

    return valid

def secondproblem(payload):
    valid = 0
    lines = [x for x in payload.split('\n') if x]

    for line in lines:
        info = line.split(' ')
        numbers = info[0].split('-')
        first = int(numbers[0])
        second = int(numbers[1])

        letter = info[1][:len(':')]

        if (info[2][first-1] == letter) != (info[2][second-1] == letter):
            valid+=1

    return valid


if __name__ == "__main__":
    dir_path = os.path.dirname(os.path.realpath(__file__))

    with open(os.path.join(dir_path, 'input'), 'r') as entry:
        payload = entry.read()

    start = time.time()
    r1 = firstproblem(payload)
    runtime_1 = time.time() - start

    start = time.time()
    r2 = secondproblem(payload)
    runtime_2 = time.time() - start

    print(f'Result: {r1} , Runtime: {runtime_1} seconds')
    print(f'Result: {r2} , Runtime: {runtime_2} seconds')

File: 2020/2/main.go
package main

import (
	"fmt"
	"go/build"
	"io/ioutil"
	"path/filepath"
	"strconv"
	"strings"
	"time"
)

func firstproblem(input string) int {
	valid := 0
	lines := strings.Split(input, "\n")

	for _, line := range lines {
		if line != "" {
			count := 0
			info := strings.Split(line, " ")
			numbers := strings.Split(info[0], "-")

			min, _ := strconv.Atoi(numbers[0])
			max, _ := strconv.Atoi(numbers[1])

			letter := info[1][:len(":")]

			for _, char := range info[2] {
				if fmt.Sprintf("%c", char) == letter {
					count++
				}
			}
			if count <= max && count >= min {
				valid++
			}
		}
	}

	return valid
}

func secondproblem(input string) int {
	valid := 0
	lines := strings.Split(input, "\n")

	for _, line := range lines {
		if line != "" {
			info := strings.Split(line, " ")
			numbers := strings.Split(info[0], "-")

			first, _ := strconv.Atoi(numbers[0])
			second, _ := strconv.Atoi(numbers[1])

			letter := info[1][:len(":")]

			if (fmt.Sprintf("%c", info[2][first-1]) == letter) != (fmt.Sprintf("%c", info[2][second-1]) == letter) {
				valid++
			}
		}
	}

	return valid
}

func main() {
	path := filepath.Join(build.Default.GOPATH, "src", "github.com", "PFadel", "adventofcode", "2020", "2", "input")

	b, err := ioutil.ReadFile(path)
	if err != nil {
		panic(err)
	}
	payload := string(b)

	start := time.Now()
	r1 := firstproblem(payload)
	elapsed1 := time.Since(start)

	start = time.Now()
	r2 := secondproblem(payload)
	elapsed2 := time.Since(start)

	fmt.Printf("%d, %f Seconds\n", r1, elapsed1.Seconds())
	fmt.Printf("%d, %f Seconds\n", r2, elapsed2.Seconds())
}
